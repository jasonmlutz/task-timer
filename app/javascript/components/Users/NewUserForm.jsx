import React, { useState } from 'react';

async function handleRegister(credentials) {
  const response = await fetch('/api/users', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'X-CSRF-TOKEN': document.querySelector('[name=csrf-token]').content,
    },
    body: JSON.stringify(credentials),
  });

  if (!response.ok) {
    const message = `An error has occurred: ${response.statusText}`;
    window.alert(message);
    return;
  }

  const user = await response.json();
  if (user.error) {
    window.alert(user.error);
  } else {
    console.log(user);
  }
}

function handleSubmit(credentials) {
  if (
    !credentials.name.length
      || !credentials.password.length
      || !credentials.password_confirmation.length
  ) {
    window.alert('please complete all fields');
  } else if (credentials.password !== credentials.password_confirmation) {
    window.alert('passwords do not match');
  } else {
    handleRegister(credentials);
  }
}

function preventSubmit(credentials) {
  return !credentials.name.length
  || !credentials.password.length
  || !(credentials.password === credentials.password_confirmation);
}

function renderPasswordStatusMessage(credentials) {
  // determine whether we need to render any message
  if (credentials.password.length
    && credentials.password_confirmation.length) {
    // set the styling and text for success or warning
    let classes = 'border-l-4 p-2 mx-2 ';
    let message;
    if (credentials.password === credentials.password_confirmation) {
      classes += 'bg-green-200 border-green-600 text-green-600';
      message = 'Passwords match!';
    } else {
      classes += 'bg-yellow-200 border-yellow-600 text-yellow-600';
      message = 'Passwords do not match!';
    }
    // return the element
    return (
      <div className="flex flex-col mb-2 mx-auto">
        <div
          className={classes}
          role="alert"
        >
          <p>{message}</p>
        </div>
      </div>
    );
  }
  // no message needed? great! return null
  return null;
}

function renderRegisterButton(credentials) {
  let classes = 'py-2 px-4 ';
  if (preventSubmit(credentials)) {
    // grayed-out theme, no mouse-over action
    classes += 'bg-gray-600 text-white w-full transition ease-in duration-200 text-center text-base font-semibold shadow-md rounded-lg cursor-not-allowed';
  } else {
    // colored and interactive
    classes += 'bg-indigo-600 hover:bg-indigo-700 focus:ring-indigo-500 focus:ring-offset-indigo-200 text-white w-full transition ease-in duration-200 text-center text-base font-semibold shadow-md focus:outline-none focus:ring-2 focus:ring-offset-2 rounded-lg';
  }
  return (
    <button
      type="submit"
      className={classes}
      disabled={preventSubmit(credentials)}
      onClick={(e) => {
        e.preventDefault();
        handleSubmit(credentials);
      }}
    >
      Register
    </button>
  );
}

function NewUserForm() {
  const [credentials, setCredentials] = useState({
    name: '',
    password: '',
    password_confirmation: '',
  });

  return (
    <div className="bg-indigo-200 relative overflow-hidden h-screen">
      <div className="container mx-auto relative z-10 mt-20">
        <div className="flex flex-col content-center relative">
          <div className="self-center mb-2 text-xl font-light text-gray-800">
            Create New Account
          </div>
          <div className="p-6 mt-8">
            <form>
              <label htmlFor="name">
                Name:
                <input type="text" id="name" className="rounded-lg flex-1 appearance-none border border-gray-300 w-full py-2 px-4 bg-white text-gray-700 placeholder-gray-400 shadow-sm text-base focus:outline-none focus:ring-2 focus:ring-indigo-600 mb-2" placeholder="please select a user name" value={credentials.name} onChange={(e) => setCredentials({ ...credentials, name: e.target.value })} />
              </label>

              <label htmlFor="password">
                Password:
                <input
                  type="password"
                  id="password"
                  className="rounded-lg flex-1 appearance-none border border-gray-300 w-full py-2 px-4 bg-white text-gray-700 placeholder-gray-400 shadow-sm text-base focus:outline-none focus:ring-2 focus:ring-indigo-600 mb-2"
                  placeholder="enter a password"
                  value={credentials.password}
                  onChange={(e) => setCredentials({ ...credentials, password: e.target.value })}
                />
              </label>
              <label htmlFor="password_confirmation">
                Confirm Password:
                <input
                  type="password"
                  id="password_confirmation"
                  className="rounded-lg flex-1 appearance-none border border-gray-300 w-full py-2 px-4 bg-white text-gray-700 placeholder-gray-400 shadow-sm text-base focus:outline-none focus:ring-2 focus:ring-indigo-600 mb-2"
                  placeholder="please re-enter your password"
                  value={credentials.password_confirmation}
                  onChange={(e) => setCredentials(
                    { ...credentials, password_confirmation: e.target.value },
                  )}
                />
              </label>
              {renderPasswordStatusMessage(credentials)}
              {renderRegisterButton(credentials)}
            </form>
          </div>
        </div>
      </div>
    </div>
  );
}

export default NewUserForm;
