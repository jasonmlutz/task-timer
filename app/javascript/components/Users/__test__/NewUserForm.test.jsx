/**
 * @jest-environment jsdom
 */
// REACT
import React from 'react';
import { createRoot } from 'react-dom/client';

// TEST UTILITIES
import { act } from 'react-dom/test-utils';
import renderer from 'react-test-renderer';
import { fireEvent, screen } from '@testing-library/react';
// eslint-disable-next-line no-unused-vars
import { toBeDisabled } from '@testing-library/jest-dom';

// COMPONENTS
import NewUserForm from '../NewUserForm';

// GLOBAL SETUP
let container;
beforeEach(() => {
  container = document.createElement('div');
  document.body.appendChild(container);
  act(() => {
    createRoot(container).render(
      <NewUserForm />,
    );
  });
});

// GLOBAL TEARDOWN
afterEach(() => {
  document.body.removeChild(container);
  container = null;
});

describe('NewUserForm component', () => {
  describe('STATIC TESTS', () => {
    test('renders correctly from snapshot', () => {
      const tree = renderer.create(<NewUserForm />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    describe('input fields are present', () => {
      test('name input', () => {
        expect(screen.getByRole('textbox', { name: /name:/i }));
      });
      test('password input', () => {
        expect(screen.getByLabelText(/^password:$/i));
      });
      test('password confirmation input', () => {
        expect(screen.getByLabelText(/^confirm password:$/i));
      });
    });

    describe('register button initial status', () => {
      test('button is present', () => {
        expect(screen.getByRole('button', { name: /register/i }));
      });
      test('button is disabled', () => {
        expect(screen.getByRole('button', { name: /register/i })).toBeDisabled();
      });
      test('button color is gray', () => {
        const button = screen.getByRole('button', { name: /register/i });
        expect(button.classList).toContain('bg-gray-600');
        expect(button.classList).not.toContain('bg-indigo-600');
      });
      test('cursor is not-available', () => {
        const button = screen.getByRole('button', { name: /register/i });
        expect(button.classList).toContain('cursor-not-allowed');
      });
    });

    describe('validation display messages are absent', () => {
      test('name availability', () => {
        expect(screen.queryAllByText(/not available/)).toStrictEqual([]);
      });
      test('password agreement', () => {
        expect(screen.queryAllByText(/do not match/)).toStrictEqual([]);
      });
    });
  });

  describe('DYNAMIC TESTS', () => {
    // LOCAL SETUP
    let nameField;
    let passwordField;
    let passwordConfirmField;
    beforeEach(() => {
      nameField = screen.getByRole('textbox', { name: /name:/i });
      passwordField = screen.getByLabelText(/^password:$/i);
      passwordConfirmField = screen.getByLabelText(/^confirm password:$/i);
    });
    describe('register button with valid field values', () => {
      let button;
      beforeEach(() => {
        button = screen.getByRole('button', { name: /register/i });
        act(() => {
          fireEvent.change(nameField, { target: { value: 'validName' } });
          fireEvent.change(passwordField, { target: { value: 'goodPassword' } });
          fireEvent.change(passwordConfirmField, {
            target: { value: 'goodPassword' },
          });
        });
      });
      test('button is enabled', () => {
        expect(button).toBeEnabled();
      });
      test('button has cursor-allowed', () => {
        expect(button.classList).not.toContain('cursor-not-allowed');
      });
      test('button is colored', () => {
        expect(button.classList).toContain('bg-indigo-600');
      });
      test('button has hover effects', () => {
        expect(button.classList).toContain('hover:bg-indigo-700');
      });
    });
    describe('password fields', () => {
      test('password fields accept input', () => {
        act(() => {
          fireEvent.change(passwordField, { target: { value: 'firstPassword' } });
          fireEvent.change(passwordConfirmField, {
            target: { value: 'secondPassword' },
          });
        });
        expect(passwordField.value).toBe('firstPassword');
        expect(passwordConfirmField.value).toBe('secondPassword');
      });
      describe('password validation message & warning', () => {
        test('non-empty non-matching passwords reveal warning', () => {
          act(() => {
            fireEvent.change(passwordField, { target: { value: 'password' } });
            fireEvent.change(passwordConfirmField, {
              target: { value: 'badPassword' },
            });
          });
          expect(screen.queryByText(/passwords do not match/i)).not.toBeNull();
        });
        test('matching passwords reveal matching message', () => {
          act(() => {
            fireEvent.change(passwordField, { target: { value: 'password' } });
            fireEvent.change(passwordConfirmField, {
              target: { value: 'password' },
            });
          });
          expect(screen.queryByText(/passwords match/i)).not.toBeNull();
        });
        test('neigher message nor warning shown when only password is empty', () => {
          act(() => {
            fireEvent.change(passwordConfirmField, {
              target: { value: 'password' },
            });
          });
          expect(screen.queryByText(/passwords do not match/i)).toBeNull();
          expect(screen.queryByText(/passwords match/i)).toBeNull();
        });
        test('neigher message nor warning shown when only password confirmation is hidden', () => {
          act(() => {
            fireEvent.change(passwordConfirmField, {
              target: { value: 'password' },
            });
          });
          expect(screen.queryByText(/passwords do not match/i)).toBeNull();
          expect(screen.queryByText(/passwords match/i)).toBeNull();
        });
      });
    });
    describe('name field', () => {
      test('name field accepts input', () => {
        act(() => {
          fireEvent.change(nameField, { target: { value: 'validName' } });
        });
        expect(nameField.value).toBe('validName');
      });
      describe('name availibility checking', () => {
        test.todo('username available message is correctly displayed');
        test.todo('username unavailable message is correctly displayed');
      });
    });
    describe('submit button action', () => {
      test.todo('triggers handleSubmit call');
      test.todo('triggers handleRegister call');
    });
  });

  describe('SUPPORTING RESOURCES TESTS', () => {
    describe('function handleRegister', () => {
      test.todo('has no return');
      test.todo('triggers a fetch API call');
      test.todo('triggers alert if !response.ok');
      test.todo('triggers alert if user.error');
      test.todo('logs user object on successful fetch');
    });

    describe('function allowSubmit', () => {
      test.todo('');
    });
  });
});
