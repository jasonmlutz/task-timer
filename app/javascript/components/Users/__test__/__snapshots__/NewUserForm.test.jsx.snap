// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewUserForm component STATIC TESTS renders correctly from snapshot 1`] = `
<div
  className="bg-indigo-200 relative overflow-hidden h-screen"
>
  <div
    className="container mx-auto relative z-10 mt-20"
  >
    <div
      className="flex flex-col content-center relative"
    >
      <div
        className="self-center mb-2 text-xl font-light text-gray-800"
      >
        Create New Account
      </div>
      <div
        className="p-6 mt-8"
      >
        <form>
          <label
            htmlFor="name"
          >
            Name:
            <input
              className="rounded-lg flex-1 appearance-none border border-gray-300 w-full py-2 px-4 bg-white text-gray-700 placeholder-gray-400 shadow-sm text-base focus:outline-none focus:ring-2 focus:ring-indigo-600 mb-2"
              id="name"
              onChange={[Function]}
              placeholder="please select a user name"
              type="text"
              value=""
            />
          </label>
          <label
            htmlFor="password"
          >
            Password:
            <input
              className="rounded-lg flex-1 appearance-none border border-gray-300 w-full py-2 px-4 bg-white text-gray-700 placeholder-gray-400 shadow-sm text-base focus:outline-none focus:ring-2 focus:ring-indigo-600 mb-2"
              id="password"
              onChange={[Function]}
              placeholder="enter a password"
              type="password"
              value=""
            />
          </label>
          <label
            htmlFor="password_confirmation"
          >
            Confirm Password:
            <input
              className="rounded-lg flex-1 appearance-none border border-gray-300 w-full py-2 px-4 bg-white text-gray-700 placeholder-gray-400 shadow-sm text-base focus:outline-none focus:ring-2 focus:ring-indigo-600 mb-2"
              id="password_confirmation"
              onChange={[Function]}
              placeholder="please re-enter your password"
              type="password"
              value=""
            />
          </label>
          <button
            className="py-2 px-4 bg-gray-600 text-white w-full transition ease-in duration-200 text-center text-base font-semibold shadow-md rounded-lg cursor-not-allowed"
            disabled={true}
            onClick={[Function]}
            type="submit"
          >
            Register
          </button>
        </form>
      </div>
    </div>
  </div>
</div>
`;
